### DISCUSSION TITLE
Header based routing

### DISCUSSION THREAD

**Discussion by evstratbg** [encode/starlette#1599] [Ideas]
Hi there,

I've needed a custom routing, based on request headers, using FastAPI. So, I implemented it, tested and now, I want to contribute to Starlette, so it can be available for everybody

Any cons?

**Comment by Kludex** [comment]
What's the use case?

**Reply by evstratbg** [reply]
API versioning, like [Stripe](https://stripe.com/docs/api/versioning) does or [Plaid](https://plaid.com/docs/api/versioning/)

**Reply by Kludex** [reply]
Did you create a separate package that implements what you want?

**Reply by evstratbg** [reply]
no, I just inherited Starlette' class and added one more method, and created another class, based on BaseRoute class. See the diff in the closed [PR](https://github.com/encode/starlette/pull/1598/files)

**Reply by evstratbg** [reply]
after this, its possible to use it like so

```
app = FastApi()

app.headers('api-version', "2020-01-01", app_1)
app.mount("/2020-01-01", app_1)
```

**Reply by evstratbg** [reply]
@Kludex any updates on this?

**Reply by adriangb** [reply]
I think this is useful functionality, but any reason why it can't be published as a standalone router?

**Reply by evstratbg** [reply]
> I think this is useful functionality, but any reason why it can't be published as a standalone router?

sorry, did I implement it wrong here? 
https://github.com/encode/starlette/pull/1598/files

**Reply by Kludex** [reply]
We are questioning the need for this functionality to be in Starlette, as it looks like  it can be a third party package.

**Reply by evstratbg** [reply]
Well, for me it looks very natural, as starlette has already a host based versioning, which is a specific case of a header based versioning. So Id rather implement BaseHeaderRouter and inherit it in Host route, than create a separate package with 1 module, actually

**Reply by Kludex** [reply]
Well, we try to maintain Starlette as minimal as possible, and as you admitdly said it can be a "package with a single module", then it will be cool if you could do that.

You're more than welcome to create a PR to add the package to our third party section on the documentation, if you think it will benefit others.

**Reply by evstratbg** [reply]
ok, got it, thank you

last thing I want to clarify. Isnt the Host based routing the part of the Header based routing? if yes, dont you think, that Header based routing should be a part of starlette?

**Reply by evstratbg** [reply]
@Kludex
